### PR TITLE
feat: [CDS-68527]: Scaling down the stage environment in BG

### PR DESCRIPTION
**Pull request by GoelPratyush99**
## Describe your changes
This PR contains the complete delegate side changes for K8s Blue Green Stage Scale Down Step. This will Scale down the workload present in stage environment. The workloads that will be scaled down are

1. DaemonSets
2. StatefulSets
3. Deployments
4. DeploymentConfigs

The workloads that will be deleted are 

1. HPA
2. PDB

All the above workload names must have the color of the environment as suffix.
The command units in this steps are

- Init (We fetch all the resources from release history depending on declarative rollback check box)
- Scale (We will scale down/delete resources  present in stage environment)

## Behaviour
CASES
1. The previous BG Deployment was Successful with Swap Services. Then we delete the workloads present in stage environment
2. The previous BG Deployment was successful but swap services have failed/ did not happen. Then we delete the the workload associated with primary service. As the blue green deployment was successful they are just a swap service away to launch their stage environment as primary.
3. The previous BG Deployment has failed and there is a successful release present in release history. Then we remove the stage environment of the earlier, if it exists.
4. The previous BG  Deployment has failed and no release is found in release history. Then we do delete/ scale down any resource.
5. The previous BG Deployment has failed and there is a successful release present in release history but it had failed at swap service so the release history was not updated. Then we scale down the workloads associated with primary service. As that deployment was successful they are just a swap service away to launch their stage environment as primary.

## Questions

1. Can there be a case specifically for custom HPA where user is point one HPA to a deployment service of specific color but the hpa name does not have the color suffix?

## Review Comment
When scale log callback is not initialised then the command unit is not executed and we get success.
Screenshot 2023-05-18 at 12 01 47 AM

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/47835)

### CODE REVIEW

**Comment by GoelPratyush99** [comment]
trigger uts

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+      if (isNotEmpty(color) && BG_SUPPORTED_COLORS.stream().anyMatch(expectedColor -> expectedColor.equals(color))) {
+        return getComplimentingColor(color);
+      }
+      return StringUtils.EMPTY;
+    }
+    throw new InvalidArgumentsException(format(
+        "Invalid subclass %s provided for %s", release.getClass().getSimpleName(), IK8sRelease.class.getSimpleName()));
+  }
+
+  private String getComplimentingColor(String color) {
```
Comment (abosii-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java:195]: There is already a method that return inverse color

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+              .k8sDelegateTaskParams(k8sDelegateTaskParams)
+              .namespace(resourceIdsToScale.get(0).getNamespace())
+              .denoteOverallSuccess(true)
+              .isErrorFrameworkEnabled(true)
+              .kubernetesConfig(kubernetesConfig)
+              .build();
+
+      K8sClient k8sClient =
+          k8sTaskHelperBase.getKubernetesClient(k8sBlueGreenStageScaleDownRequest.isUseK8sApiForSteadyStateCheck());
+      k8sClient.performSteadyStateCheck(k8sSteadyStateDTO);
```
Comment (abosii-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java:120]: We don't require steady state check

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+@Slf4j
+public class K8sBlueGreenStageScaleDownRequestHandler extends K8sRequestHandler {
+  @Inject private ContainerDeploymentDelegateBaseHelper containerDeploymentDelegateBaseHelper;
+  @Inject private K8sTaskHelperBase k8sTaskHelperBase;
+  private Kubectl client;
+  private K8sReleaseHandler releaseHandler;
+  private List<KubernetesResourceId> resourceIdsToScale;
+
+  private static final Set<String> WORKLOAD_KINDS =
+      ImmutableSet.of("Deployment", "StatefulSet", "DaemonSet", "HorizontalPodAutoscaler", "PodDisruptionBudget");
```
Comment (abosii-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java:74]: I do believe you're missing DeploymentConfig

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+  private static final Set<String> SCALE_DOWN_WORKLOAD_KINDS =
+      ImmutableSet.of("Deployment", "StatefulSet", "DaemonSet");
```
Comment (abosii-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java:76]: Here as well

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+  }
+
+  private List<KubernetesResourceId> getResourceIdsToScaleDownStageEnvironment(IK8sRelease release) {
+    if (release != null) {
+      String stageColor = getStageColor(release);
+      return release.getResourceIds()
+          .stream()
+          .filter(k8sResourceId -> WORKLOAD_KINDS.contains(k8sResourceId.getKind()) && isNotEmpty(stageColor))
+          .peek(k8sResourceId
+              -> k8sResourceId.setName(k8sResourceId.getName().replace(getComplimentingColor(stageColor), stageColor)))
```
Comment (abosii-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java:173]: This replace should handle the use case replace("abc-blue-blue", "green) -> "abc-blue-green"

File: 960-api-services/src/main/java/io/harness/k8s/releasehistory/K8sReleaseConstants.java [hunk 6]
```diff
@@ -31,4 +33,5 @@ public class K8sReleaseConstants {
   public static final int RELEASE_HISTORY_LIMIT = 2;
   public static final String RELEASE_LABEL_QUERY_SET_FORMAT = "%s in (%s)";
   public static final String RELEASE_LABEL_QUERY_LIST_FORMAT = "%s=%s";
+  public static final Set<String> BG_SUPPORTED_COLORS = ImmutableSet.of("blue", "green");
```
Comment (abosii-harness) [inline 960-api-services/src/main/java/io/harness/k8s/releasehistory/K8sReleaseConstants.java:36]: Use HarnessLabelValues.colorBlue and HarnessLabelValues.colorGreen

Comment (abosii-harness) [reply 960-api-services/src/main/java/io/harness/k8s/releasehistory/K8sReleaseConstants.java:36]: And rename it to BLUE_GREEN_COLORS

File: 930-delegate-tasks/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sScaleTaskHandler.java [hunk 7]
```diff
@@ -107,10 +108,11 @@ public K8sTaskExecutionResponse executeTaskInternal(
...
 
-    success = k8sTaskHelperBase.scale(client, k8sDelegateTaskParams, resourceIdToScale, targetReplicaCount,
-        new ExecutionLogCallback(delegateLogService, k8sScaleTaskParameters.getAccountId(),
-            k8sScaleTaskParameters.getAppId(), k8sScaleTaskParameters.getActivityId(), Scale),
-        false);
+    LogCallback scaleLogCallback = new ExecutionLogCallback(delegateLogService, k8sScaleTaskParameters.getAccountId(),
+        k8sScaleTaskParameters.getAppId(), k8sScaleTaskParameters.getActivityId(), Scale);
+    success = k8sTaskHelperBase.scale(
+        client, k8sDelegateTaskParams, resourceIdToScale, targetReplicaCount, scaleLogCallback, false);
+    scaleLogCallback.saveExecutionLog("\nDone.", INFO, SUCCESS);
```
Comment (abosii-harness) [inline 930-delegate-tasks/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sScaleTaskHandler.java:115]: you're marking it as success if if scale result is false

**Review by abosii-harness — CHANGES_REQUESTED** [review]

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+    client = Kubectl.client(k8sDelegateTaskParams.getKubectlPath(), k8sDelegateTaskParams.getKubeconfigPath());
+    IK8sReleaseHistory releaseHistory = releaseHandler.getReleaseHistory(kubernetesConfig, request.getReleaseName());
+
+    resourceIdsToScale =
+        getResourceIdsToScaleDownStageEnvironment(releaseHistory.getLatestSuccessfulBlueGreenRelease());
+    if (isEmpty(resourceIdsToScale)) {
+      executionLogCallback.saveExecutionLog(
+          "\nNo Stage environment found to scale down. Skipping the Blue Green Stage Scale Down Step", INFO);
+    } else {
+      executionLogCallback.saveExecutionLog("Manifests processed. Found following resources: \n"
```
Comment (abhinavsingh1196) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java:140]: nit: Message should be something like 
'Found following resources from stage release which are eligible for scale down:"

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+          + k8sTaskHelperBase.getResourcesIdsInTableFormat(resourceIdsToScale));
+    }
+    executionLogCallback.saveExecutionLog("\nDone.", INFO, SUCCESS);
+  }
+
+  void stageScaleDown(Kubectl client, K8sDelegateTaskParams k8sDelegateTaskParams,
+      List<KubernetesResourceId> resourceIds, LogCallback executionLogCallback) throws Exception {
+    List<KubernetesResourceId> deletedResources =
```
Comment (abhinavsingh1196) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java:148]: for readability, we can break this into 2 methods

stageScaleDown(...)
- deleteResources(resourceIds, ...)
- scaleResources(resourceIds, ...)

**Review by abhinavsingh1196 — CHANGES_REQUESTED** [review]

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java (hunk 1 above)
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java]: Done

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java (hunk 2 above)
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java]: Removed

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java (hunk 3 above)
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java]: Done

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java (hunk 4 above)
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java]: Done

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java (hunk 5 above)
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java]: Done

File: 960-api-services/src/main/java/io/harness/k8s/releasehistory/K8sReleaseConstants.java (hunk 6 above)
Comment (GoelPratyush99) [reply 960-api-services/src/main/java/io/harness/k8s/releasehistory/K8sReleaseConstants.java]: Done

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java (hunk 8 above)
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java]: Done

**Comment by GoelPratyush99** [comment]
trigger uts

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+    LogCallback logCallback =
+        k8sTaskHelperBase.getLogCallback(logStreamingTaskClient, Init, true, commandUnitsProgress);
+
+    KubernetesConfig kubernetesConfig = containerDeploymentDelegateBaseHelper.createKubernetesConfig(
+        k8sBlueGreenStageScaleDownRequest.getK8sInfraDelegateConfig(), k8sDelegateTaskParams.getWorkingDirectory(),
+        logCallback);
+    init(k8sBlueGreenStageScaleDownRequest, k8sDelegateTaskParams, kubernetesConfig, logCallback);
+
+    if (isNotEmpty(resourceIdsToScale)) {
+      LogCallback scaleLogCallback =
```
[the remark this answers is not in the document]
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java:94]: That command unit gets skipped
Screenshot 2023-05-18 at 12 01 47 AM

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    deleteStageEnvironmentResources(k8sDelegateTaskParams, executionLogCallback);
+    scaleDownStageEnvironmentResources(k8sDelegateTaskParams, executionLogCallback);
+    executionLogCallback.saveExecutionLog("\nDone.", INFO, SUCCESS);
+  }
+
+  private void deleteStageEnvironmentResources(
+      K8sDelegateTaskParams k8sDelegateTaskParams, LogCallback executionLogCallback) throws Exception {
+    List<KubernetesResourceId> deletedResources =
+        k8sTaskHelperBase.executeDeleteHandlingPartialExecution(client, k8sDelegateTaskParams,
+            resourceIdsToScale.stream()
```
Comment (abhinavsingh1196) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java:133]: this should be a separate variable in itself, method arguments like these should not be provided directly

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+            executionLogCallback, false);
+    if (isNotEmpty(deletedResources)) {
+      executionLogCallback.saveExecutionLog("Successfully deleted the following resources: \n"
+          + k8sTaskHelperBase.getResourcesIdsInTableFormat(deletedResources));
+    }
+  }
+
+  private void scaleDownStageEnvironmentResources(
+      K8sDelegateTaskParams k8sDelegateTaskParams, LogCallback executionLogCallback) throws Exception {
+    for (KubernetesResourceId resourceId : resourceIdsToScale) {
```
Comment (abhinavsingh1196) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java:145]: you could just keep using stream like above for consistency
`resourceIdsToScale.stream().filter(SCALE_WORKLOADS).forEach(scale())`

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/k8sbase/K8sTaskHelperBase.java [hunk 13]
```diff
@@ -1061,6 +1060,26 @@ private ProcessResult executeDeleteCommand(Kubectl client, K8sDelegateTaskParams
     return runK8sExecutable(k8sDelegateTaskParams, executionLogCallback, deleteCommand).getProcessResult();
   }
 
+  public boolean validateExistingResourceIds(Kubectl client, K8sDelegateTaskParams k8sDelegateTaskParams,
```
Comment (abhinavsingh1196) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/k8sbase/K8sTaskHelperBase.java:1063]: this method can be named something like `checkIfResourceExists`

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+  private List<KubernetesResourceId> getResourceIdsToScaleDownStageEnvironment(
+      IK8sRelease release, K8sDelegateTaskParams k8sDelegateTaskParams, LogCallback executionLogCallback) {
+    if (release != null) {
+      String stageColor = getStageColor(release);
+      String regex = k8sBGBaseHandler.getInverseColor(stageColor) + "$";
+      return release.getResourceIds()
+          .stream()
+          .filter(k8sResourceId -> WORKLOAD_KINDS.contains(k8sResourceId.getKind()) && isNotEmpty(stageColor))
+          .peek(k8sResourceId -> k8sResourceId.setName(k8sResourceId.getName().replaceAll(regex, stageColor)))
+          .filter(k8sResourceId
```
Comment (abhinavsingh1196) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java:162]: if you make `validateExistingResourceIds` static, then this filter predicate can be replaced with `(K8sTaskHelperBase::validateExistingResourceIds)`

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+  }
+
+  private List<KubernetesResourceId> getResourceIdsToScaleDownStageEnvironment(
+      IK8sRelease release, K8sDelegateTaskParams k8sDelegateTaskParams, LogCallback executionLogCallback) {
+    if (release != null) {
+      String stageColor = getStageColor(release);
+      String regex = k8sBGBaseHandler.getInverseColor(stageColor) + "$";
+      return release.getResourceIds()
+          .stream()
+          .filter(k8sResourceId -> WORKLOAD_KINDS.contains(k8sResourceId.getKind()) && isNotEmpty(stageColor))
```
Comment (abhinavsingh1196) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java:160]: `isNotEmpty(stageColor)` is being checked repeatedly. This method can be refactored as 
if (release == null) return emptyList()
String stageColor = getStageColor(release);

if (stageColor is empty) return emptyList()
... Rest of the method...

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+  }
+
+  private String getStageColor(IK8sRelease release) {
+    if (release instanceof K8sRelease) {
+      return k8sBGBaseHandler.getInverseColor(release.getReleaseColor());
+    }
+    if (release instanceof K8sLegacyRelease) {
+      String managedWorkloadName = ((K8sLegacyRelease) release).getManagedWorkload().getName();
+      String color = managedWorkloadName.substring(managedWorkloadName.lastIndexOf('-') + 1);
+      if (isNotEmpty(color) && BLUE_GREEN_COLORS.stream().anyMatch(expectedColor -> expectedColor.equals(color))) {
```
Comment (abhinavsingh1196) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java:177]: can use BLUE_GREEN_COLORS.contains(color)

**Review by abhinavsingh1196 — CHANGES_REQUESTED** [review]

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java (hunk 16 above)
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java]: Done

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java (hunk 15 above)
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java]: Done

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/k8sbase/K8sTaskHelperBase.java (hunk 13 above)
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/k8sbase/K8sTaskHelperBase.java]: Done

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java (hunk 11 above)
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java]: Done

File: 930-delegate-tasks/src/test/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandlerTest.java [hunk 17]
```diff
@@ -0,0 +1,368 @@
...
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.mockito.Spy;
+
+@OwnedBy(CDP)
+@FieldDefaults(level = AccessLevel.PRIVATE)
```
[the remark this answers is not in the document]
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/test/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandlerTest.java]: Done

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/k8sbase/K8sTaskHelperBase.java [hunk 18]
```diff
@@ -1061,6 +1060,26 @@ private ProcessResult executeDeleteCommand(Kubectl client, K8sDelegateTaskParams
...
+      if (result.getExitValue() == 0) {
+        return true;
+      }
+    } catch (Exception ex) {
+      log.warn("Resource {} not found in cluster. Error {}", kubernetesResourceId.kindNameRef(), ex);
+    }
+    return false;
+  }
+
+  private ProcessResult executeGetPodsCommand(Kubectl client, K8sDelegateTaskParams k8sDelegateTaskParams,
```
[the remark this answers is not in the document]
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/k8sbase/K8sTaskHelperBase.java]: Done

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java (hunk 14 above)
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sBlueGreenStageScaleDownRequestHandler.java]: I would need to change all the internal methods in K8sTaskHelperBase to static. So I am preventing that refactoring to not break existing features.

**Comment by GoelPratyush99** [comment]
trigger uts

**Review by abhinavsingh1196 — APPROVED** [review]

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
trigger review

**Pull request merged**